### ISSUE TITLE
mongoose.connect() doesn't return a promise in Angular universal, when using 'srv' in uri string

### ISSUE THREAD

**Issue by xxyyzz2050** [Automattic/mongoose#7838]
in server.ts (express server):
```
import mongoose from "mongoose"
app.get("/articles/api/*", (req, res) => {
  console.log("aaa")
   mongoose.connect(uri,{}) //uri is the connection string
       .then(db=>{
             console.log("connected")
            db.model(...) 
  })
      .catch(error=>console.log(error))

  //res.send(...)
});
```

neither .then() nor .catch() is called, but 'aaa' is called, means that the express path works fine
**note1**: running this code outside of  angular works fine
**note2**: console.log(mongoose.connect(uri)) in Angular returns zone (not promise):
`  ZoneAwarePromise { __zone_symbol__state: null, __zone_symbol__value: [] }`

when replacing mongoose.connect(..) with new Promise(r=>r()) it works fine

update:
when removing +srv from the connection string it works with error i.e: catche() runs, but with 'srv' neither .then() or .catch() runs

**Comment by vkarpov15** [comment]
Worth looking into. Do you have a sample project using angular universal?

**Comment by xxyyzz2050** [comment]
https://github.com/xxyyzz2050/eldeebCMS
@vkarpov15

**Comment by vkarpov15** [comment]
Will investigate this more. But just FYI, you shouldn't call `mongoose.connect()` in a route handler, see [this FAQ](https://mongoosejs.com/docs/faq.html#creating_connections)

**Comment by xxyyzz2050** [comment]
also error handler (and all events) never be called

**Comment by vkarpov15** [comment]
I suspect this is something weird with the mongodb server you're connecting to and/or your options parsing. I replaced your `connect()` function with one that connects to localhost:

```
export function connect(options: types.ConnectionOptions | string) {
  console.log("*** mongoose.connect() ***");
  let uri: string,
    defaultOptions = {
      useCreateIndex: true,
      useNewUrlParser: true, //https://mongoosejs.com/docs/deprecations.html; now it gives "MongoError: authentication fail"
      useFindAndModify: false,
      bufferCommands: false, //https://mongoosejs.com/docs/connections.html
      autoIndex: false,
      retryWrites: true
    };

  uri = 'mongodb://localhost:27017/test'

  console.log("uri: ", uri);
  //todo: return Promise<this mongoose, not Mongoose>

  return mongoose.connect(
    uri,
    defaultOptions
  ).then(() => console.log('Connected!'));
}
```

And now it works as expected:

```
[nodemon] 1.19.0
[nodemon] to restart at any time, enter `rs`
[nodemon] watching: *.*
[nodemon] starting `node dist/server --inspaect`
Node Express server listening on http://localhost:4200
*** mongoose.connect() ***
uri:  mongodb://localhost:27017/test
Connected!
===model===
XX: { title: 'Title', content: 'Content' }
```

**Comment by xxyyzz2050** [comment]
ok, so why id didn't reject the promise? @vkarpov15

**Comment by xxyyzz2050** [comment]
so, why it didn't reject the promise?

**Comment by xxyyzz2050** [comment]
the same uri string worked fine in an express server
I created a repo for you to try it out

https://github.com/xxyyzz2050/tmp_mongoose_issue

just run express.js and open localhost:4200

**Comment by xxyyzz2050** [comment]
is there any news? @vkarpov15 
I checked the new version 5.6.1, but the same issue still not resolved

**Comment by vkarpov15** [comment]
I'm getting an error when I run this project now:

```
$ npm start

> eldeeb-cms@0.0.1 start /home/val/Workspace/MongoDB/troubleshoot-mongoose/eldeebCMS
> npm run build  && npm run serve:server


> eldeeb-cms@0.0.1 build /home/val/Workspace/MongoDB/troubleshoot-mongoose/eldeebCMS
> npm run build:browser && npm run build:server


> eldeeb-cms@0.0.1 build:browser /home/val/Workspace/MongoDB/troubleshoot-mongoose/eldeebCMS
> ng build --prod

Unknown browser query `> 0.5 %`. Maybe you are using old Browserslist or made typo in query.
BrowserslistError: Unknown browser query `> 0.5 %`. Maybe you are using old Browserslist or made typo in query.
    at unknownQuery (/home/val/Workspace/MongoDB/troubleshoot-mongoose/eldeebCMS/node_modules/browserslist/index.js:176:10)
    at /home/val/Workspace/MongoDB/troubleshoot-mongoose/eldeebCMS/node_modules/browserslist/index.js:256:11
    at Array.reduce (<anonymous>)
    at resolve (/home/val/Workspace/MongoDB/troubleshoot-mongoose/eldeebCMS/node_modules/browserslist/index.js:197:18)
```

Realistically, I won't be able to debug this. There's simply too much code in this project for me to be able to figure out what the issue is. I can confirm Mongoose works fine with Zone.js:

```javascript
const x = require('zone.js');

console.log(global.Promise) // ZoneAwarePromise

  const mongoose = require("mongoose");
  mongoose
    .connect(
      "mongodb+srv://xxyyzz2050:Xx159753%2540%2540@cluster-test-kuwit.gcp.mongodb.net/test?retryWrites=true&w=majority"
    )
    .then(db => console.log(db))
    .catch(e => console.error(e));
```

So the issue isn't promise based, it's something in the angular config. If you can come up with a more minimal repro script, I can re-open this issue.

**Comment by xxyyzz2050** [comment]
use this repo: 
https://github.com/xxyyzz2050/angular-mongoose

read README.md & check the commits (if needed) to view how to reproduce the issue step-by-step

@vkarpov15

**Comment by vkarpov15** [comment]
I'll give this another shot :+1:

**Comment by xxyyzz2050** [comment]
many thanks for your great efforts @vkarpov15

**Comment by vkarpov15** [comment]
I think I may have found the issue :point_up: We'll see what happens with that.

But in the meantime, the only workaround I can find is not importing Zone.js. The issue comes down to Zone.js monkey-patching event emitters with cancellation logic.

**Comment by xxyyzz2050** [comment]
so, what is your recommended action to do in this example repo? @vkarpov15

**Comment by vkarpov15** [comment]
Remove Zone.js until these changes land in a new release of the mongodb node driver and Mongoose picks up those changes. Unfortunately the issue is with a non-exported variable deep in the MongoDB driver, so there's no way to work around this issue until the mongodb driver releases a fix.

**Comment by linjie0928** [comment]
I also encountered the same problem...

**Comment by juona** [comment]
@vkarpov15 

How hard would it be to determine whether an issue that I am facing is caused by the same underlying problem?

I have created a truly simple (single-file) demo repo (https://github.com/juona/mongoose-zonejs).

The problem is that when I use Mongoose in combination with Zone.js (same as @xxyyzz2050 ), and try to connect to a cluster (replica set) on Atlas, it does not work - the program simply exits. My investigation has not yielded much - it appears that the handshake requests are sent to the socket yet I do not receive any responses, after which the application decides that it has run to completion and simply exits.

This does not happen if I connect to the same cluster (same URL) using the official MongoDB driver _or_ if I do not use Zone.js.

**Comment by vkarpov15** [comment]
@juona we're still waiting on the mongodb driver to release our patch. Unfortunately there's no workaround currently besides disabling zone.js or not using Atlas

**Comment by juona** [comment]
Here's what I did to work around this problem, perhaps it will be useful to someone.

Since in my Node back-end Zone.js was only required for Angular Universal and nothing else, I only needed to find a way to run the rendering process in an isolated JavaScript environment. This could be easily achieved via a child process. Here's the basic idea in steps:

1. Collect all the code required for Angular Universal, including Zone.js, all the Angular stuff, and whatever else, and extract that to a single file.
2. Allow that process to communicate with its _parent_ via ```process.on``` and ```process.send``` in such a way that when the parent sends a "render request", this little process would perform the render and send a message back, which would contain the rendered HTML.
3. The main Node application then launches the renderer using ```fork``` from ```child_process```.
4. Finally you have to set up the application to send rendering requests  when necessary and be able to properly respond to the messages it gets back from the renderer. E.g. I am using the Express framework and basically the only thing I had to do was write an HTML engine using ```app.engine```.

I can elaborate and share pieces of code if needed, but not the whole thing as its a commercial project.

Also, disclaimer - not battle-tested, although right now I can't see why this would not work.

**Comment by vkarpov15** [comment]
This should be fixed in Mongoose 5.7.10 :+1:

**Comment by eng-dibo** [comment]
the same issue exists in mongoose 5.7.13

**Comment by vkarpov15** [comment]
@eng-dibo please open up a new issue with detailed code samples and follow the issue template. As far as I can tell, this issue is fixed.
